src/index.ts

```diff
@@ -0,0 +1,312 @@
+    queryID += "abcdefghijklmnopqrstuvwxyz0123456789".charAt(
+      Math.floor(Math.random() * 36)
+    );
+  }
+  return queryID;
+};
+
+/**
+ * TextSession creates a session based on text searches.
+ * It automatically resets once the value specified by the query label has changed in certain ways.
+ */
+export class TextSession implements ISession {
+  private queryLabel: string;
+  private session: ISession;
+  private lastQuery: string = "";
+
+  public constructor(queryLabel: string, session: ISession) {
+    this.queryLabel = queryLabel;
+    this.session = session;
+  }
+
+  public next(values: Values): [Tracking | undefined, Error | undefined] {
+    const text = values[this.queryLabel];
+    if (text === undefined) {
+      this.reset();
+      return this.session.next(values);
+    }
+
+    const shortenedPrevQ = this.lastQuery.substr(0, Math.min(text.length, 3));
+    const first3CharactersChanged = !(
+      text.substr(0, shortenedPrevQ.length) === shortenedPrevQ
+    );
+    const queryCleared = this.lastQuery.length > 0 && text.length === 0;
+    if (first3CharactersChanged || queryCleared) {
+      this.reset();
+    }
+    this.lastQuery = text;
+
+    return this.session.next(values);
+  }
+
+  public reset(): void {
+    this.session.reset();
+  }
+}
+
+export class Session implements ISession {
+  private queryID: string = "";
+  private sequence: number = 0;
+
+  private trackingType: TrackingType;
+  private field: string;
+  private sessionData: Values;
+
+  public constructor(trackingType: TrackingType, field: string, data: Values) {
+    this.trackingType = trackingType;
+    this.field = field;
+    this.sessionData = data;
+  }
+
+  public next(values: Values): [Tracking | undefined, Error | undefined] {
+    if (this.queryID === "") {
+      this.queryID = randString();
+      this.sequence = 0;
+    } else {
+      this.sequence++;
+    }
+
+    return [
+      {
+        type: this.trackingType,
+        query_id: this.queryID,
+        sequence: this.sequence,
+        field: this.field,
+        data: this.sessionData
+      },
+      undefined
+    ];
+  }
+
+  public reset(): void {
+    this.queryID = "";
+    this.sequence = 0;
+  }
+}
+
+export type Opt = (client: Client) => void;
+
+export const withEndpoint = (endpoint: string) => (client: Client) => {
+  client.endpoint = endpoint;
+};
+
+/**
+ * Client takes configuration and constructs pipelines clients
+ */
+export class Client {
+  public project: string;
+  public collection: string;
+  public endpoint: string;
+
+  public constructor(project: string, collection: string, opts: Opt[] = []) {
+    this.project = project;
+    this.collection = collection;
+    this.endpoint = "https://jsonapi.sajari.net/";
+    opts.forEach(opt => {
+      opt(this);
+    });
+  }
+
+  /**
+   * pipeline returns a new pipeline client
+   */
+  public pipeline(name: string): Pipeline {
+    return new Pipeline(this, name);
+  }
+}
+
+export interface SJError {
```

Comment:
Can this not extend `Error`? Also SJError is not a great name.

Reply:
👍